aten/src/ATen/core/autocast/autocast_mode.cpp

```diff
@@ -0,0 +1,65 @@
+#include <ATen/ATen.h>
+#include <ATen/core/autocast/autocast_mode.h>
+#include <c10/core/impl/LocalDispatchKeySet.h>
+
+#include <stdexcept>
+#include <memory>
+
+namespace at {
+namespace autocast {
+
+/// thread_local is a feature that is not enabled by Caffe2 mobile
+/// build (e.g. iOS). Therefore, we only provide `at::AutocastMode`
+/// when we are not in mobile build or when FEATURE_TORCH_MOBILE
+/// is on.
+#if !defined(C10_MOBILE) || defined(FEATURE_TORCH_MOBILE)
```

Comment:
I think the macros here are unnecessary; the implementation of `tls_is_dispatch_key_included` itself will fall back to a plain old global variable (non thread local) if thread local is not supported. So you can avoid having to define all the stubs here.

Reply:
Making the flag global changes the behavior in a potentially unsafe way, firstly because accesses to the flag are not mutexed, and secondly because if a bunch of threads are enabling/disabling autocast for their own purposes but writing to the same global flag when they do so, one of them might disable it while another one wants/expects it to be enabled.

Unless it's a guarantee that the mobile build is executed single-threadedly, I'd rather be conservative and error out.  What do you think?

Reply:
The assumption is that it's executed only single threadedly. There is *a lot* of stuff that won't work if that assumption is not upheld, you aren't the only thread local in town ;)

Reply:
sounds good to me, shared guilt is easier to bear.  Does that mean I don't need the macros at all anywhere in my PR (specifically, I don't need them [here](https://github.com/pytorch/pytorch/pull/32140#discussion_r389908847) either)?

Reply:
yes, assuming that it builds